File: crates/klio-diagnostics/src/render.rs
//! Diagnostic renderers. Each format consumes the same `[Diagnostic]` slice
//! against a `SourceMap`; choose the renderer that matches your downstream
//! consumer (terminal, JSON-consuming tooling, SARIF aggregator).

pub mod json;
pub mod plain;
pub mod sarif;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Format {
    Plain,
    Json,
    Sarif,
}

impl std::str::FromStr for Format {
    type Err = ();

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        Ok(match s {
            "plain" => Self::Plain,
            "json" => Self::Json,
            "sarif" => Self::Sarif,
            _ => return Err(()),
        })
    }
}

File: src/diagnostics/render.zig
//! Diagnostic renderers. Each format consumes the same `[Diagnostic]` slice
//! against a `SourceMap`; choose the renderer that matches your downstream
//! consumer (terminal, JSON-consuming tooling, SARIF aggregator).

const std = @import("std");

pub const plain = @import("render/plain.zig");
pub const json = @import("render/json.zig");
pub const sarif = @import("render/sarif.zig");

pub const Format = enum {
    Plain,
    Json,
    Sarif,

    pub fn fromStr(s: []const u8) ?Format {
        if (std.mem.eql(u8, s, "plain")) return .Plain;
        if (std.mem.eql(u8, s, "json")) return .Json;
        if (std.mem.eql(u8, s, "sarif")) return .Sarif;
        return null;
    }
};

test "format from str" {
    try std.testing.expectEqual(Format.Plain, Format.fromStr("plain").?);
    try std.testing.expectEqual(Format.Json, Format.fromStr("json").?);
    try std.testing.expectEqual(Format.Sarif, Format.fromStr("sarif").?);
    try std.testing.expect(Format.fromStr("nope") == null);
}
